File: src/Homeworks/Homework2.Tests/ParserTests.cs
using System.Collections.Generic;
using Homework2.IL;
using Xunit;

namespace Homework2.Tests
{
	public class ParserTests
	{
		[Theory]
		[MemberData(nameof(CorrectArguments))]
		public void TryParseArguments_CorrectExpression_ReturnZero(string[] args)
		{
			var actual = Parser.TryParseArguments(args, out _, out _, out _);
			Assert.Equal(0, actual);
		}

		public static IEnumerable<object[]> CorrectArguments()
		{
			yield return new object[] {new[] {"10", "+", "2"}};
			yield return new object[] {new[] {"2", "-", "10"}};
			yield return new object[] {new[] {"20", "*", "10"}};
			yield return new object[] {new[] {"4", "/", "4"}};
			yield return new object[] {new[] {"50", ":", "2"}};
		}

		[Theory]
		[MemberData(nameof(InvalidArguments))]
		public void TryParseArguments_InvalidArguments_ReturnOne(string[] args)
		{
			var actual = Parser.TryParseArguments(args, out _, out _, out _);
			Assert.Equal(1, actual);
		}

		public static IEnumerable<object[]> InvalidArguments()
		{
			yield return new object[] {new[] {"x", "-", "2"}};
			yield return new object[] {new[] {"5", "+", "="}};
			yield return new object[] {new[] {"p", "+", "x"}};
		}

		[Theory]
		[MemberData(nameof(InvalidOperations))]
		public void TryParseArguments_InvalidOperation_ReturnTwo(string[] args)
		{
			var actual = Parser.TryParseArguments(args, out _, out _, out _);
			Assert.Equal(2, actual);
		}

		public static IEnumerable<object[]> InvalidOperations()
		{
			yield return new object[] {new[] {"10", "<", "2"}};
			yield return new object[] {new[] {"5", "\\", "5"}};
			yield return new object[] {new[] {"10", "%", "6"}};
		}
	}
}

File: src/Homeworks/Homework4.Tests/ParserTests.fs
module Tests.Parser

open Homework4
open Xunit

let checkParsingOperation args expected =
    let mutable val1 = 0
    let mutable val2 = 0
    let mutable operation = ""
    let actual = Parser.TryParseArguments args &val1 &operation &val2
    Assert.Equal(expected, actual)

let CorrectArguments = [|
    [|[|"10"; "+"; "2"|]|]
    [|[|"2"; "-"; "10"|]|]
    [|[|"20"; "*"; "10"|]|]
    [|[|"4"; "/"; "4"|]|]
    [|[|"50"; ":"; "2"|]|]
|]

[<Theory>]
[<MemberData(nameof(CorrectArguments))>]
let ``TryParseArgments_CorrectExpression_ReturnZero`` (args) =
    checkParsingOperation args 0

let InvalidArguments = [|
    [|[|"x"; "-"; "2"|]|]
    [|[|"5"; "+"; "="|]|]
    [|[|"p"; "+"; "x"|]|]
|]

[<Theory>]
[<MemberData(nameof(InvalidArguments))>]
let ``TryParseArguments_InvalidArguments_ReturnOne`` (args) =
    checkParsingOperation args 1

let InvalidOperations = [|
    [|[|"10"; "<"; "2"|]|]
    [|[|"5"; "\\"; "5"|]|]
    [|[|"10"; "%"; "6"|]|]
|]

[<Theory>]
[<MemberData(nameof(InvalidOperations))>]
let ``TryParseArguments_InvalidOperation_ReturnTwo`` (args) =
    checkParsingOperation args 2